### ISSUE TITLE
CPU frequency control fails for CPU 4

### ISSUE THREAD

**Issue by qzed** [linux-surface/surface-pro-x#14]
CPU frequency control seems to fail for CPU 4 with the following log: 

```
[    1.984565] cpu cpu4: Voltage update failed freq=3072000
[    1.984571] cpu cpu4: failed to update OPP for freq=3072000
[    1.984576] cpu cpu4: Voltage update failed freq=3148800
[    1.984579] cpu cpu4: failed to update OPP for freq=3148800
[    1.984582] cpu cpu4: Voltage update failed freq=3148800
[    1.984585] cpu cpu4: failed to update OPP for freq=3148800
```

In addition to this, CPU core 0 seems to be pinned to ~80% usage.

**Comment by AquaBx** [comment]
Adding this to the DTS seems to mitigate the issue. I will try benching the CPU before and after. 

```
&cpu4_opp_table {
	opp-3072000000 {
		opp-hz = /bits/ 64 <3072000000>;
		opp-peak-kBps = <8368000 51609600>;
	};
	
	opp-3148800000 {
		opp-hz = /bits/ 64 <3148800000>;
		opp-peak-kBps = <8368000 51609600>;
	};
};
```

Update:

|                                     | Before                     | After                      | Speedup      |
|-------------------------------------|----------------------------|----------------------------|--------------|
| sysbench run cpu                    | 3355.68 events per second  | 3540.83 events per second  | x1.055175106 |
| sysbench --threads=$(nproc) run cpu | 16682.78 events per second | 17371.16 events per second | x1.041262907 |
| Geekbench Singlecore  | 914| 952 | x1.041575492 |
| Geekbench Multicore |3099 | 3256 | x1.050661504 |

https://browser.geekbench.com/v6/cpu/compare/15897297?baseline=15897749

**Comment by AquaBx** [comment]
I tried to increase the speed, but the score improved by a little (https://browser.geekbench.com/v6/cpu/15916735)

```
&cpu4_opp_table {

	opp-3072000000 {
		opp-hz = /bits/ 64 <3072000000>;
		opp-peak-kBps = <68256000 68256000>; # 32 * 2133000
	};
	
	opp-3148800000 {
		opp-hz = /bits/ 64 <3148800000>;
		opp-peak-kBps = <68256000 68256000>;
	};
};
```

**Comment by qzed** [comment]
Just out of curiosity, how did you get these values?

**Comment by AquaBx** [comment]
First, I have copied the ones from the sc8180x.dtsi for lower frequencies.

I have then checked the ram speed on Windows to try to increase the speed

https://github.com/linux-surface/kernel/blob/1cc181c914274d50fcb69c7c8d7db9a691811f2c/arch/arm64/boot/dts/qcom/sc8180x.dtsi#L502-L505

**Comment by AquaBx** [comment]
Also why 2133000*32

I have seen in the sc8280x that they were using this calculation, so I give it a shot with higher speed.
https://github.com/linux-surface/kernel/blob/1cc181c914274d50fcb69c7c8d7db9a691811f2c/arch/arm64/boot/dts/qcom/sc8280xp.dtsi#L568C1-L571C5

But honestly, I don't know if they are worth it. More digging on that I guess.

**Comment by qzed** [comment]
Hmm, I would probably go with your first values as I'm a bit concerned whether the increased ones are stable across all devices.

**Comment by AquaBx** [comment]
I agree. My math is purely theoretical and I don't know if the CPU handles it properly.
At least it hasn't crashed on my SQ2, but the gains are between margins of errors.

**Comment by qzed** [comment]
Will be included in the next release.
